designs/nodepool-node-limits.md

```diff
@@ -0,0 +1,143 @@
+# NodePool Node Limits in Karpenter
+
+## Motivation
+
+Service providers and cluster administrators should have the ability to enforce maximum node counts per `NodePool` to address several critical use cases:
+
+* **Licensing constraints**: Respect software licensing limits (e.g., monitoring agents with per-node licensing)
+* **IP address management**: With fixed IP address pools, physically prevent node creation beyond the available IP capacity
+* **Service Provider SLAs**: Guarantee service level agreements by preventing unlimited node provisioning
+* **Resource limits**: Respect cloud provider account limits or organizational quotas with Karpenter-level control rather than relying on infrastructure-level restrictions
+
+## Notes
+
+- Enforcing global maximum Node limits is not a goal of this design.
+- This design does not involve nodes managed outside of Karpenter.
+- The Node limits discussed in this design are different from the Node limits discussed in the [Karpenter - Static Capacity](static-capacity.md) design, but involve similar concepts.
+    - See the [Relationship to Static Capacity Design](#relationship-to-static-capacity-design) section for more details.
+
+## Design Options
+
+There are three potential approaches for implementing `NodePool` node limits, each with different trade-offs:
+
+### Option 1: Hard Limits
+
+**Description**: Node limits are enforced as hard constraints that block both provisioning and disruption when the limit is reached. Disruption is blocked because Karpenter requires the new nodes to be schedulable and ready before the old node can be terminated, which can lead to more nodes than the limit at one state.
```

Comment:
I think its worth calling out that not all disruption requires new node replacement. For example, empty or simply consolidating into existing nodes. So the final implementation may be turning off paths that may cause a new node launch while letting some disruption to still occur.

In addition, Expiration is "Forceful" where it doesn't wait for a replacement to be launch before terminating a node. 

Is locking both provision and disruption ideal? I am wondering if there is something else that Karpenter does instead